docs/lang/articles/advanced/layout.md

```diff
@@ -132,177 +109,183 @@ for (int i = 0; i < 3; i++) {
+<!-- For example, the following places two 1D fields of shape `3` together, which
+is called Array of Structures (AoS):
 
 ```python
-for i, j in A:
-    A[i, j] += 1
+ti.root.dense(ti.i, 3).place(x, y)
 ```
 
-will iterate over elements of `A` following the row-major order. Similarly, if
-`A` is defined in column-major, then the iteration follows the column-major
-order.
+Their memory layout is:
 
-If `A` is hierarchical, it will be iterated level by level. This maximizes the
-memory bandwidth utilization in most cases.
+By contrast, the following places these two fields separately, which is called
+Structure of Arrays (SoA): -->
 
-As you may notice, only dense data layouts are covered in this section. For sparse
-data layouts, see [Sparse computation](./sparse.md).
+<!-- ```python
+ti.root.dense(ti.i, 3).place(x)
+ti.root.dense(ti.i, 3).place(y)
+```
 
-### More examples of advanced dense data layouts
+Now their memory layout is:
 
-2D field, row-major:
 
-```python
-A = ti.field(ti.f32)
-ti.root.dense(ti.ij, (256, 256)).place(A)
-```
+**To improve spatial locality of memory accesses, it may be helpful to
+place data elements that are often accessed together within
+relatively close addresses.**  -->
 
-2D field, column-major:
-```python
-A = ti.field(ti.f32)
-ti.root.dense(ti.j, 256).dense(ti.i, 256).place(A)
-```
 
-_8x8_-blocked 2D field of size _1024x1024_:
+### AoS extension: hierarchical fields
+<!-- haidong: I hope to remove this subsection. This content just repeats the AoS topic -->
+Sometimes we want to access memory in a complex but fixed pattern, like traversing an image in 8x8 blocks. The apparent best practice is to flatten each 8x8 block and concatenate them together. From a Taichi user's view, however, the field is no longer a flat buffer. It now has a hierarchy with two levels: the image level and the block level. Equivalently, the field is an array of implicit 8x8 block strucutrues.
+
+We demonstrate the statements as follows:
 
 ```python
-density = ti.field(ti.f32)
-ti.root.dense(ti.ij, (128, 128)).dense(ti.ij, (8, 8)).place(density)
+# Flat field
+val = ti.field(ti.f32)
+ti.root.dense(ti.ij, (M, N)).place(val)
 ```
 
-3D particle positions and velocities, AoS:
-
 ```python
-pos = ti.Vector.field(3, dtype=ti.f32)
-vel = ti.Vector.field(3, dtype=ti.f32)
-ti.root.dense(ti.i, 1024).place(pos, vel)
-# equivalent to
-ti.root.dense(ti.i, 1024).place(pos.get_scalar_field(0),
-                                pos.get_scalar_field(1),
-                                pos.get_scalar_field(2),
-                                vel.get_scalar_field(0),
-                                vel.get_scalar_field(1),
-                                vel.get_scalar_field(2))
+# Hierarchical field
+val = ti.field(ti.f32)
+ti.root.dense(ti.ij, (M // 8, N // 8)).dense(ti.ij, (8, 8)).place(val)
 ```
+where `M` and `N` are multiples of 8. We hope the readers try on their own compute machines. The performance difference can be amazing.
```

Comment:
```suggestion
where `M` and `N` are multiples of 8. We encourage users to try this out! The performance difference can be significant!
```

Reply:
Always address your readers "you". Avoid "users".

Reply:
Good advice!